### PR TITLE
Extended workshop subscriber

### PR DESCRIPTION
**Pull request by Revadike**
Logs and counts errors + skips unneeded workshop items, as requested by @tfedor

### CODE REVIEW

File: js/content/community.js [hunk 1]
```diff
@@ -3670,29 +3671,53 @@ let WorkshopBrowseClass = (function(){
...
                     }
 
-                    window.dialog = ShowBlockingWaitDialog("${Localization.str[method + "_all"]}", \`${Localization.str[method + "_loading"].replace("__i__", ++i).replace("__count__", total)}\`);
+                    window.dialog = ShowBlockingWaitDialog("${Localization.str[method + "_all"]}", \`${Localization.str[method + "_loading"].replace("__i__", completed).replace("__count__", total)}${failed ? ` (${Localization.str.failed.replace("__n__", failed)})` : ""}\`);
                 }`)
             }
 
             function changeSubscription(id) {
-                return new Promise(function(resolve) {
+                return new Promise(function(resolve, reject) {
```
Comment (KarlCastle) [inline js/content/community.js]: Why `new Promise()`? Why not use the Promise returned by `RequestData.post()`?

File: js/content/community.js [hunk 2]
```diff
@@ -3702,14 +3727,22 @@ let WorkshopBrowseClass = (function(){
                     let result = await RequestData.getHttp(url.toString());
```
Comment (KarlCastle) [inline js/content/community.js]: It would be nice if we also caught any errors that getHttp might throw.

File: js/content/community.js [hunk 3]
```diff
@@ -3670,42 +3671,75 @@ let WorkshopBrowseClass = (function(){
...
+                let formData = new FormData();
+                formData.append("sessionid", User.getSessionId());
+                formData.append("appid", appid);
+                formData.append("id", id);
+
+                return RequestData.post("https://steamcommunity.com/sharedfiles/" + method, formData, {
+                    withCredentials: true
+                }, true)
+                .then(function(res) {
+                    if (!res || !res.success || res.success !== 1) {
```
Comment (MxtOUT) [inline js/content/community.js]:
```suggestion
                    if (!res || !res.success) {
```
That is already checked in the previous condition

File: js/content/community.js [hunk 4]
```diff
@@ -3670,42 +3671,75 @@ let WorkshopBrowseClass = (function(){
...
+                        failed++;
+                        console.error(new Error("Bad response"))
+                    }
+                    completed++;
+                    updateWaitDialog();
```
Comment (MxtOUT) [inline js/content/community.js]: You might as well put this and the line above into a `finally` callback

File: js/content/community.js [hunk 5]
```diff
@@ -3670,42 +3671,75 @@ let WorkshopBrowseClass = (function(){
...
+                formData.append("appid", appid);
+                formData.append("id", id);
+
+                return RequestData.post("https://steamcommunity.com/sharedfiles/" + method, formData, {
+                    withCredentials: true
+                }, true)
+                .then(function(res) {
+                    if (!res || !res.success || res.success !== 1) {
+                        failed++;
+                        console.error(new Error("Bad response"))
```
Comment (MxtOUT) [inline js/content/community.js]: You can `throw` here, the `catch` callback will handle this

File: js/content/community.js [hunk 6]
```diff
@@ -3670,42 +3671,75 @@ let WorkshopBrowseClass = (function(){
...
                 for (let p = 1; p <= Math.ceil(total / 30); p++) {
                     let url = new URL(window.location.href);
                     url.searchParams.set("p", p);
                     url.searchParams.set("numperpage", 30);
     
-                    let result = await RequestData.getHttp(url.toString());
-                    let xmlDoc = new DOMParser().parseFromString(result, "text/html");
+                    let result = await RequestData.getHttp(url.toString()).catch(err => console.error(err));
+                    if (!result) {
+                        await sleep(10000);
```
Comment (MxtOUT) [inline js/content/community.js]: I wouldn't retry in these cases, I'd log and continue

File: js/content/community.js [hunk 7]
```diff
@@ -3670,42 +3671,75 @@ let WorkshopBrowseClass = (function(){
...
+                        p--;
+                        continue;
+                    }
 
-                    for (let node of xmlDoc.querySelectorAll(".workshopItemPreviewHolder")) {
+                    let xmlDoc = new DOMParser().parseFromString(result, "text/html");
```
Comment (MxtOUT) [inline js/content/community.js]: Do we have to create a new `DOMParser` instance for every page?

File: js/content/community.js [hunk 8]
```diff
@@ -3670,42 +3671,75 @@ let WorkshopBrowseClass = (function(){
...
+                    
+                        node = node.querySelector(".workshopItemPreviewHolder");
                         workshopItems.push(node.id.replace("sharedfile_", ""))
                     }
                 }
+
+                total = workshopItems.length;
+                updateWaitDialog();
     
                 Promise.all(
```
Comment (MxtOUT) [inline js/content/community.js]: I know that this isn't part of this PR, but the formatting of this statement suggest that `finally` is called on the result of `Array.map`, although it is called on `Promise.all`

File: js/content/community.js (hunk 3 above)
Comment (Revadike) [reply js/content/community.js]: I believe the success code could be NOT 1, in which case it it wrong.

Comment (MxtOUT) [reply js/content/community.js]: I guess this is meant to be a boolean value, so there's either 0 or 1

File: js/content/community.js (hunk 8 above)
Comment (Revadike) [reply js/content/community.js]: This formatting was done by @tfedor. I suggest you discuss this with him.

File: js/content/community.js (hunk 4 above)
Comment (Revadike) [reply js/content/community.js]: Resolved in https://github.com/tfedor/AugmentedSteam/pull/506/commits/30c9ee0eb891468bf90becf0dc44a7d9dd19657a

File: js/content/community.js (hunk 5 above)
Comment (Revadike) [reply js/content/community.js]: Resolved in https://github.com/tfedor/AugmentedSteam/pull/506/commits/0c3fd463f0c96a4ad74c16e14489c4e2f6baa767

File: js/content/community.js (hunk 6 above)
Comment (Revadike) [reply js/content/community.js]: Resolved in https://github.com/tfedor/AugmentedSteam/pull/506/commits/f58d20a58f438b82ee7a9e257498a535f161c4f2

File: js/content/community.js (hunk 7 above)
Comment (Revadike) [reply js/content/community.js]: Resolved in https://github.com/tfedor/AugmentedSteam/pull/506/commits/19578cf351a6b620e97d2a55738daff0f61413dd

**Pull request merged**